modules/libcom/src/fdmgr/fdManager.cpp

```diff
@@ -79,62 +147,115 @@ LIBCOM_API fdManager::~fdManager()
 //
 LIBCOM_API void fdManager::process (double delay)
 {
-    this->lazyInitTimerQueue ();
+    priv->lazyInitTimerQueue ();
 
     //
     // no recursion
     //
-    if (this->processInProg) {
+    if (priv->processInProg)
         return;
-    }
-    this->processInProg = true;
+    priv->processInProg = true;
 
     //
     // One shot at expired timers prior to going into
-    // select. This allows zero delay timers to arm
+    // poll/select. This allows zero delay timers to arm
     // fd writes. We will never process the timer queue
     // more than once here so that fd activity get serviced
     // in a reasonable length of time.
     //
-    double minDelay = this->pTimerQueue->process(epicsTime::getCurrent());
+    double minDelay = priv->pTimerQueue->process(epicsTime::getCurrent());
 
     if ( minDelay >= delay ) {
         minDelay = delay;
     }
 
-    bool ioPending = false;
-    tsDLIter < fdReg > iter = this->regList.firstIter ();
+#ifdef FDMGR_USE_POLL
+    priv->pollfds.clear();
+#endif
+
+    int ioPending = 0;
+    tsDLIter<fdReg> iter = priv->regList.firstIter();
     while ( iter.valid () ) {
-        FD_SET(iter->getFD(), &this->fdSetsPtr[iter->getType()]);
-        ioPending = true;
+        ++ioPending;
+
+#ifdef FDMGR_USE_POLL
+#if __cplusplus >= 201100L
+        priv->pollfds.emplace_back(pollfd{.fd = iter->getFD(), .events = PollEvents[iter->getType()]});
+#else
+        struct pollfd pollfd;
+        pollfd.fd = iter->getFD();
+        pollfd.events = PollEvents[iter->getType()];
+        pollfd.revents = 0;
+        priv->pollfds.push_back(pollfd);
+#endif
+#endif
+
+#ifdef FDMGR_USE_SELECT
+        FD_SET(iter->getFD(), &priv->fdSets[iter->getType()]);
+#endif
         ++iter;
     }
 
     if ( ioPending ) {
+#ifdef FDMGR_USE_POLL
+        if (minDelay * mSecPerSec > INT_MAX)
+            minDelay = INT_MAX / mSecPerSec;
+
+        int status = poll(&priv->pollfds.front(), // ancient C++ has no vector.data()
```

Comment:
wrt. "ancient C++".  The pattern I have seen used most is `&vec[0]`.

More significantly, be careful as either one is UB if `vec.empty()`.  The MSVC library has `assert()` that this is so.

Is it true that `ioPending!=0` implies `!priv->pollfds.empty()`?